File: src/lib/api/insights.ts
import { apiRequest } from "./client";

export interface TodayScheduleItem {
  id: number;
  subject: string;
  start_time: string;
  end_time: string;
  room?: string | null;
  class_id: number;
  class_name: string;
  teacher_name?: string | null;
}

export interface TodayWorkItem {
  kind: "assignment" | "quiz";
  id: number;
  title: string;
  subject?: string | null;
  deadline: string;
  is_overdue: boolean;
  class_id: number;
  class_name: string;
  action_url: string;
}

export interface AttentionItem {
  id: string;
  kind: "sos" | "missing_assignment" | "low_quiz_score";
  priority: "high" | "medium";
  title: string;
  description: string;
  class_id: number;
  class_name: string;
  student_id?: number | null;
  action_url: string;
  occurred_at?: string | null;
}

export interface TodayDashboard {
  date: string;
  schedule: TodayScheduleItem[];
  work_items: TodayWorkItem[];
  attention: AttentionItem[];
  unread_notifications: number;
}

export interface GradebookStudent {
  student_id: number;
  student_name: string;
  student_email: string;
  assignment_average?: number | null;
  quiz_average?: number | null;
  overall_average?: number | null;
  graded_items: number;
  total_items: number;
  missing_items: number;
  needs_attention: boolean;
}

export interface ClassGradebook {
  class_id: number;
  class_name: string;
  assignment_count: number;
  quiz_count: number;
  students: GradebookStudent[];
  pagination: {
    page: number;
    page_size: number;
    total_items: number;
    total_pages: number;
  };
}

export interface StudentSubjectGrade {
  subject: string;
  assignment_average?: number | null;
  quiz_average?: number | null;
  overall_average?: number | null;
  completed_items: number;
  graded_items: number;
  total_items: number;
  needs_review: boolean;
}

export interface StudentGradebook {
  class_id?: number | null;
  class_name?: string | null;
  overall_average?: number | null;
  subjects: StudentSubjectGrade[];
}

export const insightsApi = {
  getToday: () => apiRequest<TodayDashboard>("/api/dashboard/today", { cache: "no-store" }),
  getClassGradebook: (classId: number, page = 1, pageSize = 100) =>
    apiRequest<ClassGradebook>(
      `/api/classes/${classId}/gradebook?page=${page}&page_size=${pageSize}`,
      { cache: "no-store" },
    ),
  getStudentGradebook: () =>
    apiRequest<StudentGradebook>("/api/student/gradebook", { cache: "no-store" }),
};

File: backend/app/schemas/insights.py
"""Typed contracts for small, factual school-workflow aggregates."""

from typing import Literal

from pydantic import BaseModel, Field


class TodayScheduleItem(BaseModel):
    id: int
    subject: str
    start_time: str
    end_time: str
    room: str | None = None
    class_id: int
    class_name: str
    teacher_name: str | None = None


class TodayWorkItem(BaseModel):
    kind: Literal["assignment", "quiz"]
    id: int
    title: str
    subject: str | None = None
    deadline: str
    is_overdue: bool
    class_id: int
    class_name: str
    action_url: str


class AttentionItem(BaseModel):
    id: str
    kind: Literal["sos", "missing_assignment", "low_quiz_score"]
    priority: Literal["high", "medium"]
    title: str
    description: str
    class_id: int
    class_name: str
    student_id: int | None = None
    action_url: str
    occurred_at: str | None = None


class TodayDashboardResponse(BaseModel):
    date: str
    schedule: list[TodayScheduleItem]
    work_items: list[TodayWorkItem]
    attention: list[AttentionItem]
    unread_notifications: int = 0


class GradebookStudentRow(BaseModel):
    student_id: int
    student_name: str
    student_email: str
    assignment_average: float | None = None
    quiz_average: float | None = None
    overall_average: float | None = None
    graded_items: int = 0
    total_items: int = 0
    missing_items: int = 0
    needs_attention: bool = False


class Pagination(BaseModel):
    page: int = Field(ge=1)
    page_size: int = Field(ge=1, le=100)
    total_items: int = Field(ge=0)
    total_pages: int = Field(ge=1)


class ClassGradebookResponse(BaseModel):
    class_id: int
    class_name: str
    assignment_count: int
    quiz_count: int
    students: list[GradebookStudentRow]
    pagination: Pagination


class StudentSubjectGrade(BaseModel):
    subject: str
    assignment_average: float | None = None
    quiz_average: float | None = None
    overall_average: float | None = None
    completed_items: int = 0
    graded_items: int = 0
    total_items: int = 0
    needs_review: bool = False


class StudentGradebookResponse(BaseModel):
    class_id: int | None = None
    class_name: str | None = None
    overall_average: float | None = None
    subjects: list[StudentSubjectGrade]
